docs/guides/filesystem-usage.md

```diff
@@ -0,0 +1,191 @@
+# How to use the lakeFS file system
+
+This guide contains instructions and code snippets on how to use the lakeFS file system.
+
+## The lakeFS URI structure
+
+In the following subsections, we frequently make use of [lakeFS URIs](https://docs.lakefs.io/understand/model.html#lakefs-protocol-uris) in the example code.
+lakeFS URIs identify resources in a lakeFS deployment through a unique path consisting of repository name, lakeFS revision/ref name, and file name relative to the repository root.
+
+As an example, a URI like `repo/main/file.txt` addresses the `file.txt` file on the `main` branch in the repository named `repo`.
+
+In some lakeFS file system operations, directories are also allowed as resource names.
+For example, the URI `repo/main/data/` (note the trailing slash) refers to the `data` directory on the `main` branch in the `repo` repository.
+
+## On staged versus committed changes
+
+When uploading, copying, or removing files or directories from a branch, those removal operations will result in staged changes in the repository until a commit is created.
+`lakefs-spec` does not create these commits automatically, since it separates file operations from versioning operations rigorously.
+If you want to conduct versioning operations, like creating commits, between file transfers, the best way to do so is by using [filesystem transactions](transactions.md).
+
+## How to use lakeFS file system APIs
+
+The following section explains more in-depth how to use the `LakeFSFileSystem` APIs.
+This section concerns the explicitly implemented operations. In addition, there are a number of file system APIs inherited from the [`AbstractFileSystem` interface in fsspec](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.spec.AbstractFileSystem).
+
+More information on file system usage can be found in the [`fsspec` documentation](https://filesystem-spec.readthedocs.io/en/latest/usage.html#use-a-file-system).
+
+### Uploading and downloading files
+
+The arguably most important feature of the file system is file transfers.
+
+#### File uploads
+
+To upload a file, you can use the `fs.put()` and `fs.put_file()` methods. 
+While `fs.put_file()` operates on single files only, the `fs.put()` API can be used for directory uploads.
+
+```python
+from lakefs_spec import LakeFSFileSystem
+
+fs = LakeFSFileSystem()
+
+# remote path, then local target path.
+fs.put_file("file.txt", "my-repo/my-ref/file.txt")
+```
+
+If you want to upload an entire directory to lakeFS, you can use the `fs.put()` API together with the `recursive=True` switch:
+
+```python
+# structure:
+#   dir/a.txt
+#      /b.yaml
+#      /c.csv
+#      /...
+
+fs.put("dir", "my-repo/my-ref/dir", recursive=True)
```

Comment:
Having tried this, does not actually work, but rather raises an error:

```
Traceback (most recent call last):
  File "/work/lakefs-spec/.sandbox/demo.py", line 17, in <module>
    fs.put("data", f"{REPO}/{BRANCH}/data/", recursive=True)
  File "/work/lakefs-spec/.direnv/python-3.11/lib/python3.11/site-packages/fsspec/spec.py", line 1069, in put
    self.put_file(lpath, rpath, **kwargs)
  File "/work/lakefs-spec/src/lakefs_spec/spec.py", line 437, in put_file
    local_checksum = md5_checksum(lpath, blocksize=self.blocksize)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work/lakefs-spec/src/lakefs_spec/util.py", line 11, in md5_checksum
    with open(lpath, "rb") as f:
         ^^^^^^^^^^^^^^^^^
IsADirectoryError: [Errno 21] Is a directory: '/work/lakefs-spec/data'
```

The root cause is that the internal implementation resolves the local path (`fs.expand_path(lpath, recursive=recursive, maxdepth=maxdepth)` in `AbstractFileSystem.put()`), but that includes the directory itself, not just the files contained therein.

The below seems to do the trick (note that the `**` glob runs into the same problem): 


```suggestion
fs.put("dir/*", "my-repo/my-ref/dir", recursive=True)
```

However, this might be worth following up on, because it makes for some surprising behavior when compared to regular globs.

Reply:
I'll change the occurrences in this file to something that works, but let's still follow up on this later on.

Reply:
Addressed in 7e09b41c9f7d3ac7a0b2582832432358ebf4ff42, which makes changing the example unnecessary.